cli/listener/fuzz/README.md

```diff
@@ -0,0 +1,30 @@
+# APY Daemon fuzzer
+
+The APY daemon is internet-exposed and accepts user input, so we should fuzz it
+to find any inputs that might trigger a panic (for example, overflow in the
+datetime handling, etc.).
+
+First, install `cargo-fuzz`:
+
+    cargo +nightly-2022-03-22 install cargo-fuzz --vers 0.11.0
+
+Then run a (slow) fuzzer in debug mode with address sanitizer, from the
+`cli/listener` directory:
+
+    cargo +nightly-2022-03-22 fuzz run apy_endpoint \
+      --jobs=«num-cores» \
+      -- -dict=fuzz/dictionary.txt
+
+Or run a faster fuzzer but with less intricate coverage:
+
+    cargo +nightly-2022-03-22 fuzz run apy_endpoint \
+      --jobs=«num-cores» \
+      --release \
+      --sanitizer=none \
+      --debug-assertions \
+      -- -dict=fuzz/dictionary.txt
```

Comment:
> This feature is experimental, right?

No it’s not experimental, it has been there for a long time. Note that the arguments after the `--` are for libFuzzer and not for `cargo-fuzz`, you can run it with `-- -help=1` to learn more. About dictionaries, [the docs are here](https://llvm.org/docs/LibFuzzer.html#dictionaries).

> Wonder which fields it tries to use the dict for

It doesn’t try “fields”, libFuzzer only generates bytestrings. It’s the `arbitrary` crate that generates struct instances from the bytestring. It does it in such a way that the mutations that libFuzzer applies tend to create, usually correspond to interesting mutations in the generated struct. You can think of `Arbitrary` as a specialized deserializer.

libFuzzer takes the dictionary into account in the mutations that it applies to the bytestring. It will sometimes splice entries from the dictionary into the bytestring, but it still has all of its other mutations, the dictionary just makes it faster. For example, for the string "begin", how would libFuzzer discover it? It would try random inputs until it hits a “b” and the `memcmp` gets one step further. That’s a new program path so this input is interesting, we mutate it further, and at some point it will hit the “e”, etc. But it will take `128 * len("begin")` iterations on average, if it was mutating at the right place, and it gets worse for longer strings, so we can kickstart it a bit with a dictionary. But even without the dictionary, it would eventually discover those paths.